<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Bacterial Genome Assembly: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="manifest" href="../site.webmanifest">
<link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><abbr class="badge badge-light" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="background-color: #FF4955; border-radius: 5px">
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #000">
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
            Pre-Alpha
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Bacterial Genome Assembly
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Bacterial Genome Assembly
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<hr>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a class="btn btn-primary" href="../aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Bacterial Genome Assembly
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->
      
            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Bacterial Genome Assembly using Trycycler</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="demo.html">2. Evaluating Clusters</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="reconcile.html">3. Reconciling clusters</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width resources">
<a href="../instructor/aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">
            
            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-introduction"><p>Content from <a href="introduction.html">Bacterial Genome Assembly using Trycycler</a></p>
<hr>
<p>Last updated on 2024-04-19 |
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/introduction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 32 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why use Trycycler to assemble bacterial genomes with long-read
data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand common errors that can occur during assembly</li>
<li>Become familiar with the Trycycler approach</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="introduction"><h2 class="section-heading">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>Bacterial genomes are dynamic landscapes. They may frequently lose
and gain both chromosomal and extrachromosomal elements such as
insertion sequences (IS), REP sequences, ICE elements, plasmids and
prophage.</p>
<p>This can be particularly interesting to food microbiologists in the
context of adaptations to dynamic niches - such as during food
fermentations! In order to detect and characterise this kind of genomic
plasticity, we require high quality bacterial genome assemblies which
are both:</p>
<ul>
<li>Complete - one contig per replicon</li>
<li>Accurate - fully match the actual DNA sequence of the organism</li>
</ul>
<p>Short-read sequencing (e.g. Illumina) generates highly accurate
reads, but the short length of reads (typically 150-300bp) cannot
resolve repetitive or difficult-to-sequence regions, resulting in
fragmented draft assemblies. Importantly, dynamic genome elements such
as IS, REP and ICE are often themselves repetitive, meaning that their
presence is not reflected in the draft assembly, yet they can be quite
influential on the phenotype of bacterial isolates.</p>
<div id="insertion-sequences-is" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="insertion-sequences-is" class="callout-inner">
<h3 class="callout-title">Insertion sequences (IS)<a class="anchor" aria-label="anchor" href="#insertion-sequences-is"></a>
</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show more</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Insertion sequences (IS) are short genomic elements which are
frequently found in bacterial genomes. They can shape genomes as they
move around, potentially introducing deletions, mutations and
influencing gene expression. <img src="https://www.ncbi.nlm.nih.gov/pmc/articles/instance/9241782/bin/spectrum.02112-21-f007.jpg" alt="Insertion Sequences" class="figure"></p>
</div>
</div>
</div>
</div>
<div id="rep-sequences" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="rep-sequences" class="callout-inner">
<h3 class="callout-title">REP sequences<a class="anchor" aria-label="anchor" href="#rep-sequences"></a>
</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show more</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>REP sequences are genomic regions containing highly repetitive and
palindromic sequences. They are often located in the the extragenic
space of some bacterial genomes e.g. ERIC (Enterobacterial Repetitive
Intergenic Consensus) sequences in the Enterobacterales. <img src="https://d3i71xaburhd42.cloudfront.net/9d601ea51726a3257a80ba2f7cc9f98c4a569397/2-Figure1-1.png" alt="Examples of rep sequences" class="figure"></p>
</div>
</div>
</div>
</div>
<div id="integrative-and-conjugative-elements-ices" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="integrative-and-conjugative-elements-ices" class="callout-inner">
<h3 class="callout-title">Integrative and Conjugative Elements
(ICEs)<a class="anchor" aria-label="anchor" href="#integrative-and-conjugative-elements-ices"></a>
</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show more</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p>Integrative and Conjugative Elements (ICEs) are mobile genetic
elements capable of moving between hosts and donors through conjugation.
They are flanked on either side by direct repeat sequences, which allow
them to excise via site-specific recombination. <img src="https://media.springernature.com/lw685/springer-static/image/art%3A10.1007%2Fs10142-022-00903-2/MediaObjects/10142_2022_903_Fig2_HTML.png?as=webp" alt="ICEs" class="figure"></p>
</div>
</div>
</div>
</div>
</section><section id="assembling-the-perfect-bacterial-genome"><h2 class="section-heading">Assembling the “perfect” bacterial genome?<a class="anchor" aria-label="anchor" href="#assembling-the-perfect-bacterial-genome"></a>
</h2>
<hr class="half-width">
<p>Bacterial genome assembly remains an open problem - Torsten Seeman,
immortalises this concept in the stdout of his assembly pipeline
Shovill: “Remember, an assembly is just a <strong>hypothesis</strong> of
the original sequences”!</p>
<p>Hybrid bacterial whole-genome assembly combines the accuracy of short
reads with the additional information provided by long reads generated
on Oxford Nanopore Technologies (ONT) platforms. Previous approaches
used a “short-read-first” approach, only using long reads to connect
short contigs generated from Illumina data. Improvements in yield and
accuracy of ONT data has caused a gradual shift in the field to
“long-read-first” pipelines, using short reads only for the polishing of
long-read assemblies.</p>
<div id="example-long-read-first-assembly-pipeline" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="example-long-read-first-assembly-pipeline" class="callout-inner">
<h3 class="callout-title">Example: long-read-first assembly
pipeline:<a class="anchor" aria-label="anchor" href="#example-long-read-first-assembly-pipeline"></a>
</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">Show me the solution</h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p>Note: this is by far from the only possible pipeline and does not
include, for example, PacBio sequencing. <img src="https://www.researchgate.net/publication/368938787/figure/fig1/AS:11431281124501087@1677968652785/Illustrated-overview-of-our-recommended-approach-to-perfect-bacterial-whole-genome.png" alt="Assembling the perfect bacterial genome using Oxford Nanopore and Illumina sequencing" class="figure"></p>
</div>
</div>
</div>
</div>
<p>Other pipelines exist which completely automate the long-read
assembly process, such as Hybracter (<a href="https://github.com/gbouras13/hybracter" class="external-link uri">https://github.com/gbouras13/hybracter</a>) or Dragonflye
(<a href="https://github.com/rpetit3/dragonflye" class="external-link uri">https://github.com/rpetit3/dragonflye</a>). Trycycler on the
other hand is deterministic and requires manual intervention at
different points. So, one of the obvious <code>questions</code> could
be: why dedicate extra time to using Trycycler?</p>
<div id="closed-genomes-versus-accuracy" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="closed-genomes-versus-accuracy" class="callout-inner">
<h3 class="callout-title">Closed genomes versus accuracy<a class="anchor" aria-label="anchor" href="#closed-genomes-versus-accuracy"></a>
</h3>
<div class="callout-content">
<p>Flye assembled one circular contig for my bacterial chromosome -
isn’t that good enough?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5">Show more</h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<p>It is possible to assemble a bacterial genome and find that it is
“closed” as one circular chromosome - however it may not be very
accurate or in fact complete! Long-read-only assemblies can contain a
variety of errors, such as “indels” (deletions ~50bp), or missassemblies
where contigs are joined in incorrect orientations.</p>
<figure><img src="https://journals.plos.org/ploscompbiol/article/figure/image?size=large&amp;id=10.1371/journal.pcbi.1010905.g002" alt="Assembling the perfect bacterial genome using Oxford Nanopore and Illumina sequencing" class="figure mx-auto d-block"><div class="figcaption">Assembling the perfect bacterial genome using
Oxford Nanopore and Illumina sequencing - reproduced from
10.1371/journal.pcbi.1010905</div>
</figure>
</div>
</div>
</div>
</div>
</section><section id="overview-of-trycycler"><h2 class="section-heading">Overview of Trycycler<a class="anchor" aria-label="anchor" href="#overview-of-trycycler"></a>
</h2>
<hr class="half-width">
<p>Trycycler is a tool developed by Ryan Wick and colleagues to address
an issue they had uncovered in previous benchmarking work. Several
long-read assembly tools exist for genome assembly, each with their own
strengths and tendency to generate certain assembly errors, for
example:</p>
<ul>
<li>Flye: fast, accurate, has dedicated plasmid setting, high RAM
usage</li>
<li>Raven: fast, low RAM usage, issues with circularisation and missing
small plasmids</li>
<li>Miniasm (within Unicycler): good at achieving circularisation, but
not the best at completing chromosomal contigs</li>
<li>Canu: slow, good at recovering plasmids, tendency to produce
chimeric replicons + hinder circularisation</li>
</ul>
<p>Keeping with the idea that any assembly is a
<strong>hypothesis</strong> of the original bacterial genome, Trycycler
looks to improve the strength of our <strong>hypothesis</strong> by
finding a consensus amongst different assemblers and subsets of the
entire read set. In doing so, we should be better able to identify and
correct large-scale assembly errors.</p>
<figure><img src="https://www.researchgate.net/publication/354593531/figure/fig1/AS:1068171633106963@1631683371578/Overview-of-the-Trycycler-long-read-assembly-pipeline-Before-Trycycler-is-run-the-user.png" alt="Overview of the Trycycler long-read assembly pipeline" class="figure mx-auto d-block"><div class="figcaption">Overview of the Trycycler long-read assembly
pipeline - reproduced from Wick et al (2021)</div>
</figure><p>It is important to remember that Trycycler assemblies are still
unlikely to be perfect, and will frequently contain “homopolymer errors”
- however these can be corrected by long-read and particularly
short-read polishing (there is debate on how much polishing is helpful -
for further reading see Ryan Wick’s blog post: <a href="https://rrwick.github.io/2023/11/06/accuracy-vs-depth-update.html" class="external-link uri">https://rrwick.github.io/2023/11/06/accuracy-vs-depth-update.html</a>).</p>
<div id="a-more-detailed-look-at-the-trycycler-process" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="a-more-detailed-look-at-the-trycycler-process" class="callout-inner">
<h3 class="callout-title">A more detailed look at the Trycycler
process<a class="anchor" aria-label="anchor" href="#a-more-detailed-look-at-the-trycycler-process"></a>
</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6">Show me the solution</h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<figure><img src="https://github.com/rrwick/Trycycler/wiki/images/pipeline.png" alt="Illustrated pipeline overview" class="figure mx-auto d-block"><div class="figcaption">Illustrated pipeline overview - reproduced from
<a href="https://github.com/rrwick/Trycycler/wiki/" class="external-link uri">https://github.com/rrwick/Trycycler/wiki/</a>
</div>
</figure>
</div>
</div>
</div>
</div>
</section><section id="enough-theory-time-for-a-demo"><h2 class="section-heading">Enough theory, time for a demo!<a class="anchor" aria-label="anchor" href="#enough-theory-time-for-a-demo"></a>
</h2>
<hr class="half-width">
<p>Now that we have had a whistle-stop tour of bacterial genome
assebmly, we will use Trycycler on some real world data!</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Illumina short-read sequencing is accurate but cannot produce
complete bacterial genomes</li>
<li>ONT long-read sequencing is less accurate, but can resolve gaps and
repeats</li>
<li>No assembler is perfect, but Trycycler lets you find the best
consensus</li>
<li>Long-read-first assemblies can be improved with short-read
polishing</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-demo"><p>Content from <a href="demo.html">Evaluating Clusters</a></p>
<hr>
<p>Last updated on 2024-04-21 |
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/demo.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 32 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can we navigate the Migale cluster to evaluate clusters?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Become familiar with the directory structure and commands on
Migale/SGE</li>
<li>Understand how to evaluate and mark “bad” clusters</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="introduction"><h2 class="section-heading">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>Short (Illumina NovaSeq) and long (ONT P2solo) read data has been
generated for a selection of bacterial isolates. After quality control,
Trycycler’s “subsample” command was used to generate 3 subsamples of
each readset at a minimum allowed read depth of 20x, and providing the
expected genome size. Each subsample was then assembled by each
assembler (Flye, Raven and Unicycler), generating a total of 9
assemblies. (One assembly was made for each (not-subsampled) readset
using Canu - it wasn’t included in the Trycycler clustering but could be
useful later for troubleshooting more challenging assemblies).</p>
<p>Step 2 of Trycycler (<a href="https://github.com/rrwick/Trycycler/wiki/Clustering-contigs" class="external-link">Clustering
contigs</a>) has already been run, so let’s take a look at our starting
data.</p>
<pre><code><span> <span class="va">cd</span> <span class="op">/</span><span class="va">save_projet</span><span class="op">/</span><span class="va">domino_wp3_isollates</span></span></code></pre>
<p>Each sample directory should have a sub-directory containing the
input assemblies,a sub-directory called ‘trycycler’ and a .fastq.gz
containing the reads for that sample.</p>
<figure><img src="../fig/folder1.jpg" alt="Sample directory structure" class="figure mx-auto d-block"><div class="figcaption">Sample directory structure</div>
</figure><p>Before starting, we will copy the sample folder to the working
directory:</p>
<pre><code><span><span class="va">rsync</span> <span class="op">-</span><span class="va">av</span> <span class="op">/</span><span class="va">save_projet</span><span class="op">/</span><span class="va">domino_wp3_isollates</span><span class="op">/</span><span class="va">test_CIRM_BIA1910</span> <span class="op">/</span><span class="va">work_projet</span><span class="op">/</span><span class="va">domino_wp3_isollates</span><span class="op">/</span><span class="va">test_CIRM_BIA1910</span></span>
<span><span class="va">cd</span> <span class="op">/</span><span class="va">work_projet</span><span class="op">/</span><span class="va">domino_wp3_isollates</span><span class="op">/</span><span class="va">test_CIRM_BIA1910</span></span></code></pre>
<p>When we descend into the ‘trycycler’ folder we can see the output
from Trycycler clustering, with one folder for each cluster, and .newick
and .phylip files which contain information on the clustering of contigs
(complete-linkage clustering based on Mash distance).</p>
<figure><img src="../fig/folder2.jpg" alt="Trycyler directory structure" class="figure mx-auto d-block"><div class="figcaption">Trycycler directory structure</div>
</figure><p>At the start, each cluster directory contains just one directory
called “1_contigs” which contains the .fasta sequences of all contigs
assigned to that cluster:</p>
<figure><img src="../fig/folder3.jpg" alt="Cluster directory structure" class="figure mx-auto d-block"><div class="figcaption">Cluster directory structure</div>
</figure><div id="callout1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout1"></a>
</h3>
<div class="callout-content">
<p>Trycycler renames contigs with letters (“A”, “B”, “C”…). to simplify
the visualisation, but you can still identify which assembler produced
each one. Contig names containing “Utg” were generated by Raven; Flye
contigs contain “contig_X” where X is a number; Unicycler contigs simply
contain “_X”.</p>
</div>
</div>
</div>
</section><section id="cluster-assessment"><h2 class="section-heading">Cluster assessment<a class="anchor" aria-label="anchor" href="#cluster-assessment"></a>
</h2>
<hr class="half-width">
<p>While it is possible to run the next step (Reconciling contigs)
automatically on all clusters, it is likely to fail on many clusters.
This is the most manual part and subjective part of the Trycycler
process, but will improve the quality of our final assemblies.</p>
<p>The Trycycler wiki goes into greater detail on what to consider when
evaluating clusters <a href="https://github.com/rrwick/Trycycler/wiki/Clustering-contigs#choose-your-clusters" class="external-link">here</a>.
In an ideal world (!), sequencing an <em>E. coli</em> strain carrying
one plasmid would produce two clusters: cluster_001 generally being the
chromosomal cluster, and cluster_002 containing the plasmid contigs.</p>
<p>Additionally, if all assemblers performed equally well on all read
subsets, we would expect to see 9 contigs in cluster_001, each of which
roughly in alignment with our expected genome size, and 9 contigs in
cluster_002 with about the same length. This can happen on occasion, but
more often you will need to intervene to exclude erroneous clusters.
Start by checking the number of clusters Trycyler has found.</p>
<p>A genome could have more than 10 valid clusters if it were carrying
many plasmids, but generally &gt;10 clusters indicates that either the
sequencing wasn’t deep enough, or that there was a lot of heterogeneity
in the reads. Check the spreadsheet to see if there were a very low or
high number of reads/bases generated for that sample. The mean/median
read length is also important - too short reads will be unable to span
repetitive regions and can cause fragmentation or misassembly.
Ultimately, we won’t be able to do much if the reads are too few/short,
but if the sample was sequenced to very high depth, we might be able to
generate better assemblies by subsetting further or trying a different
assembler.</p>
<p>Next, view the tree. You can either copy and paste the .newick file
into an in-browser tool like <a href="http://etetoolkit.org/treeview/" class="external-link">ETEToolkit</a>, or use the
software installed on Migale with the supplied helper script in
‘/save_projet/domino_wp3_isollates/scripts/ete.py’</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="im">from</span> ete3 <span class="im">import</span> Tree</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co"># Need to include this for Qt bug</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>os.environ[<span class="st">'QT_QPA_PLATFORM'</span>]<span class="op">=</span><span class="st">'offscreen'</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co"># Read the tree</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>t <span class="op">=</span> Tree(<span class="st">'trycycler/contigs.newick'</span>)</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co"># Render the tree to a PNG file</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>t.render(<span class="st">"contigs_newick.png"</span>)</span></code></pre>
</div>
<p>You can execute this in the sample directory with this command:</p>
<pre><code>qsub -cwd -V -N tree_view -o out/tree_view.o -e err/tree_view.e -b y "conda activate ete3-3.1.3 &amp;&amp; python3 /save_projet/domino_wp3_isollates/scripts/ete.py &amp;&amp; conda deactivate"</code></pre>
<p>It should produce “contigs_newick.png” in the trycycler folder, which
you can then open and view. <img src="../fig/contigs_newick.png" alt="Contigs_newick png generated with ETEToolkit" class="figure"></p>
<p>To remove a cluster, you can simply rename the folder e.g. to
rm_cluster_002 (or bad_cluster_002 - it doesn’t matter once the
directory name does not commence with “cluster_”. This keeps the
sequences, in case we want to revisit the decision later on.</p>
<figure><img src="../fig/folder4.jpg" alt='Marking a "bad" cluster' class="figure mx-auto d-block"><div class="figcaption">Marking a “bad” cluster directory by
renaming</div>
</figure><div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div>Hackathon: semi-automation of cluster assessment?</h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler1" aria-labelledby="headingSpoiler1">
<div class="accordion-body">
<p>If you can, make a quick note of why you decided to exclude/include
clusters on the recording sheet.</p>
<p>It is unlikely that Trycycler can be fully automated, but we could
consider writing rules to automate the discarding of obvious poor
clusters, e.g.</p>
<ul>
<li>if the number of contigs in a cluster is &lt;2, discard</li>
<li>if a cluster is highly-related to a chromomosomal cluster and
contains &lt;2, discard</li>
<li>when is it worth keeping “fragments” which could describe true <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9980784/" class="external-link">structural
heterogeneity</a> e.g. (pro)phage, <em>fim</em> switches, conjugative
plasmids?</li>
<li>etc!</li>
</ul>
</div>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-reconcile"><p>Content from <a href="reconcile.html">Reconciling clusters</a></p>
<hr>
<p>Last updated on 2024-04-21 |
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/reconcile.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 32 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What happens next to “good” clusters we identified?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand how and why to exclude a contig from a cluster</li>
<li>Be able to manually repair duplicated contigs</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="introduction"><h2 class="section-heading">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>In the previous step we identified clusters of contigs, which could
each represent true replicons present in our isolates. However, they
were primarily clustered using Mash distance and before alignment, they
need to be standardised in terms of length, similarity and orientation.
We will use Trycycler’s ‘reconcile’ function to achieve this. (<a href="https://github.com/rrwick/Trycycler/wiki/Reconciling-contigs" class="external-link">Reconciling
contigs</a>). Additionally, as most (but not all!) bacterial replicons
are circular, Trycycler will try to cleanly circularise each contig
using the other other contigs in the cluster as a reference: for more
information, see: <a href="https://github.com/rrwick/Trycycler/wiki/How-circularisation-repair-works" class="external-link">Trycycler
wiki: How circularisation repair works</a>.</p>
<figure><img src="../fig/folder4.jpg" alt="Trycycler folder after clusters have been manually assessed" class="figure mx-auto d-block"><div class="figcaption">Trycycler folder after clusters have been
manually assessed</div>
</figure></section><section id="running-trycycler-reconcile"><h2 class="section-heading">Running Trycycler Reconcile<a class="anchor" aria-label="anchor" href="#running-trycycler-reconcile"></a>
</h2>
<hr class="half-width">
<p>The basic command to reconcile a cluster looks like this:</p>
<pre><code>trycycler reconcile --reads *.fastq.gz --cluster_dir trycycler/cluster_001 </code></pre>
<p>and we could submit this as a cluster job using the following command
(changing the number of the cluster each time for the stderr, stdout and
–cluster_dir parameters):</p>
<pre><code>qsub -cwd -V -N tr_test -e err/try_rec_001 -o out/try_rec_001 -pe thread 4 -b y "conda activate trycycler-0.5.4 &amp;&amp; trycycler reconcile --reads *.fastq.gz --cluster_dir trycycler/cluster_001 --threads 4 &amp;&amp; conda deactivate"
qsub -cwd -V -N tr_test -e err/try_rec_002 -o out/try_rec_002 -pe thread 4 -b y "conda activate trycycler-0.5.4 &amp;&amp; trycycler reconcile --reads *.fastq.gz --cluster_dir trycycler/cluster_002 --threads 4 &amp;&amp; conda deactivate"</code></pre>
<p>However, for simplicity, a bash script is provided which will detect
the clusters present, and generate and submit a script to run Trycycler
reconcile on each. You can run it like this, with no arguments (in the
main isolate directory):</p>
<pre><code>/save_projet/domino_wp3_isollates/scripts/reconcile.sh</code></pre>
<p>If Trycycler was able to reconcile a cluster automatically, a new
file ‘2_all_seqs.fasta’ will be generated in the cluster directory,
containing the reconciled contigs. Frequently, it will fail to produce
output, meaning that manual intervention is required to resolve. The
error file (written to err/try_rec_{clusternumber}) will guide your next
course of action.</p>
</section><section id="manual-interventions"><h2 class="section-heading">Manual interventions<a class="anchor" aria-label="anchor" href="#manual-interventions"></a>
</h2>
<hr class="half-width">
<p>Reconciliation can fail for many reasons <a href="https://github.com/rrwick/Trycycler/wiki/Reconciling-contigs#manual-intervention" class="external-link">Reconciling
contigs - manualintervention</a>.</p>
<p>Most commonly, one or more contigs in a cluster will be too long or
short. Mash distance greater than the default threshold (0.01) can also
prevent the reconciliation from finishing. Like we saw before with the
cluster folders, we can exclude a contig by renaming its extension from
‘contig.fasta’ to ‘contig.bad’, or similiar. You can then rerun
Trycycler reconcile on just that cluster:</p>
<pre><code>qsub -cwd -V -N tr_test -e err/try_rec_001 -o out/try_rec_001 -pe thread 4 -b y "conda activate trycycler-0.5.4 &amp;&amp; trycycler reconcile --reads *.fastq.gz --cluster_dir trycycler/cluster_001 --threads 4 &amp;&amp; conda deactivate"</code></pre>
<p>If we are left with very few contigs in a cluster (e.g. &lt;3) then
it will probably be worth making more assemblies before proceeding
further - as we started with 9 assemblies, we have some leeway.</p>
</section><section id="manual-repair-of-overlap"><h2 class="section-heading">Manual repair of overlap<a class="anchor" aria-label="anchor" href="#manual-repair-of-overlap"></a>
</h2>
<hr class="half-width">
<p>You may encounter a cluster with a contig that is longer than the
others, preventing reconciliation. This can indicate an assembly error
and is particularly common with plasmids, which can be double or even
tripled in length. We can manually remove the (presumed) duplicate
sequence in a text editor: <a href="https://github.com/rrwick/Trycycler/wiki/FAQ-and-miscellaneous-tips#manually-fixing-overlap" class="external-link">Trycycler
Wiki: Manually fixing overlap</a>.</p>
</section><section id="linear-plasmidsphage"><h2 class="section-heading">Linear plasmids/phage<a class="anchor" aria-label="anchor" href="#linear-plasmidsphage"></a>
</h2>
<hr class="half-width">
<p>Phage genomes can adopt linear conformations. Linear plasmids have
also been described in some bacteria in the order ‘Enterobacterales’
e.g. <em>Klebsiella</em>, <em>Salmonella</em>. If you have have a
suspicious cluster of small contigs that will not circularise, you could
try to <a href="https://www.ncbi.nlm.nih.gov/geo/query/blast.html" class="external-link">Blast</a> one
of the sequences and see the closest hits. More frequently, such
clusters tend to represent chromosomal fragments e.g. prophage, which
one or more assemblers excluded from the main chromosome.</p>
</section><section id="modifying-the-parameters-for-trycycler-reconcile"><h2 class="section-heading">Modifying the parameters for Trycycler reconcile<a class="anchor" aria-label="anchor" href="#modifying-the-parameters-for-trycycler-reconcile"></a>
</h2>
<hr class="half-width">
<p>The default parameters of the ‘reconcile’ function work well for most
cases, and it is likely preferrable to exclude a handful of contigs
first. In certain scenarios, you may wish to amend the <a href="https://github.com/rrwick/Trycycler/wiki/Reconciling-contigs#settings" class="external-link">thresholds</a>,
particularly ‘–max_length_diff’, ‘–max_add_seq’,
‘–max_add_seq_percent’,‘–max_trim_seq’ and ‘–max_trim_seq_percent’. You
can add the flags to your ‘qsub’ submission command.</p>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/README.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries/workbench-template-rmd/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:katie.fala@teagasc.ie">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.4" class="external-link">sandpaper (0.16.4)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.5" class="external-link">pegboard (0.7.5)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.2" class="external-link">varnish (1.0.2)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries.github.io/workbench-template-rmd/instructor/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "genomics, assembly, bacterial, trycycler",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://carpentries.github.io/workbench-template-rmd/instructor/aio.html",
  "identifier": "https://carpentries.github.io/workbench-template-rmd/instructor/aio.html",
  "dateCreated": "2024-04-18",
  "dateModified": "2024-04-21",
  "datePublished": "2024-04-21"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

